db/fielddef.go

```diff
@@ -0,0 +1,153 @@
+package db
+
+import (
+	"database/sql"
+
+	"github.com/filecoin-project/boost/storagemarket/types/dealcheckpoints"
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/big"
+	"github.com/filecoin-project/go-state-types/crypto"
+	"github.com/ipfs/go-cid"
+	"golang.org/x/xerrors"
+)
+
+type fieldDefinition interface {
+	fieldPtr() interface{}
+	marshall() (interface{}, error)
+	unmarshall() error
+}
+
+type fieldDef struct {
+	f interface{}
+}
+
+var _ fieldDefinition = (*fieldDef)(nil)
+
+func (fd *fieldDef) fieldPtr() interface{} {
+	return fd.f
+}
+
+func (fd *fieldDef) marshall() (interface{}, error) {
+	return fd.f, nil
+}
+
+func (fd *fieldDef) unmarshall() error {
+	return nil
+}
+
+type cidFieldDef struct {
+	cidStr sql.NullString
+	f      *cid.Cid
+}
+
+func (fd *cidFieldDef) fieldPtr() interface{} {
+	return &fd.cidStr
+}
+
+func (fd *cidFieldDef) marshall() (interface{}, error) {
+	if fd.f == nil {
+		return nil, nil
+	}
+	return fd.f.String(), nil
+}
+
+func (fd *cidFieldDef) unmarshall() error {
+	if !fd.cidStr.Valid {
+		return nil
+	}
+
+	c, err := cid.Parse(fd.cidStr.String)
+	if err != nil {
+		return xerrors.Errorf("parsing CID from string '%s': %w", fd.cidStr, err)
+	}
+
+	*fd.f = c
+	return nil
+}
+
+type bigIntFieldDef struct {
+	marshalled []byte
+	f          *big.Int
+}
+
+func (fd *bigIntFieldDef) fieldPtr() interface{} {
```

Comment:
Good catch 🎣 
The database type should be BLOB

Reply:
@dirkmc - Do you know why the test did NOT catch this ?

Reply:
Because sqllite is very flexible with types, so it doesn't actually enforce types on insert / update, it just sets the row to whatever type it gets: https://www.sqlite.org/datatype3.html#affinity